### ISSUE TITLE
HMR works with Svelte but not React/Preact/Solid integration

### ISSUE THREAD

**Issue by kidonng** [withastro/astro#6742]
### What version of `astro` are you using?

2.1.9

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

I've been trying out (and lately, settled on) Astro since before v1.0, and HMR in Astro has went through several changes. One thing bugging me now is that, HMR, in the sense that reloading framework components doesn't require a page refresh, works with Svelte integration, but not React, Preact, or Solid. I haven't tried other integrations so maybe some of them works.

Little is mentioned in the Astro docs, while I expect them to work since they do work in an ordinary Vite project. Is this a bug? Or just not implemented? Or there are some limitations?

### Link to Minimal Reproducible Example

https://astro.new/

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by natemoo-re** [comment]
There is a difference between Hot Module Reloading (HMR) and **Fast Refresh** which preserves component state and swaps modules without requiring a full page refresh. You are noticing that the latter does not work for JSX frameworks, but does work for other frameworks like Vue and Svelte.

This is a limitation of our current architecture. Astro currently has to implement JSX transforms internally rather than relying on the official P/React/Solid Vite plugins (which also provide Fast Refresh transforms) because each of those plugins assumes that every J/TSX file in a project can be considered as P/React/Solid component. In Astro, that is not the case. For example, it is possible to have some J/TSX files that use Preact and some that use Solid.

In the future, we would like to ease this restriction. But we'd ideally find a way to do that without taking on the burden of implementing our own fast refresh logic that diverges from the official Vite plugins.

**Comment by kidonng** [comment]
Thanks for the background!

> because each of those plugins assumes that every J/TSX file in a project can be considered as P/React/Solid component

Those plugins all provide a `include` option ([Preact](https://github.com/preactjs/preset-vite/blob/main/src/index.ts#L40-L43), [React](https://github.com/vitejs/vite-plugin-react/tree/main/packages/plugin-react#includeexclude), [Solid](https://github.com/solidjs/vite-plugin-solid#optionsinclude)), could that be useful? I know they are not dynamic though.

Or, if the user only configured a single JSX renderer, then use the official Vite plugin directly?

**Comment by natemoo-re** [comment]
Nope, unfortunately the `include` option isn't very useful because it only matches filenames, not the content of the file. Astro JSX support relies on matching _the contents_ of a file to a specific framework.

I think configuring a single renderer _and_ using the official Vite plugin could work, but I wouldn't be surprised if that breaks some internal things.

How to unblock this is definitely an open question—we've been discussing it internally for a long time. Unfortunately I'm not sure where it fits on the roadmap at the moment.

**Comment by matthewp** [comment]
A lot of people have asked for React Refresh support. I do think we could do something like, if there's only 1 JSX library in a project, just use the Vite plugins so this problem doesn't exist.

In any case, there's nothing actionable to be done in this issue. Please open a Roadmap issue about support React Refresh, I would support it myself!
